src/services/fetch/useData.ts

```diff
@@ -0,0 +1,22 @@
+import { useEffect, useState } from "react";
+
+export const useData = <T>(url: string | null) => {
+  const [data, setData] = useState<T[] | null>(null);
+
+  useEffect(() => {
+    if (url !== null && url !== "null") {
+      let ignore = false;
+      fetch(url)
+        .then((response) => response.json())
+        .then((json) => {
+          if (!ignore) {
+            setData(json);
+          }
+        });
+      return () => {
+        ignore = true;
+      };
+    }
+  }, [url]);
+  return data;
+};
```

Comment:
aqui ficou um pouco confuso, tem ifs dentro de ifs e fica dificil para uma outra pessoa que for utilizar entender o que acontece, existe uma coisa chamda "early return" para evitar if nested. Mas creio que você deva refatorar para cobrir as partes de `getCountries` que comentei você pode criar como funções mesmo` const getCountries = async () => ...` e depois exportar cada uma das funções` return { getCountries, ...}`

Reply:
qualquer duvida pode chamar também